### ISSUE TITLE
Use knitr::purl() to export Python files (.py) 

### ISSUE THREAD

**Issue by GitHunter0** [yihui/knitr#1928]
Hey @yihui , purl() has being an incredibly useful function for me. 

Since RStudio editor for .py files is still limited (no code sections and no outline pane), I mostly use R Notebooks to write Python code.

Is it possible to make purl() to export .py files as well, so that we can later source it via reticulate::source_python()? 

It would be very handy.

Thank you very much

**Comment by yihui** [comment]
Yes, I hope to improve this in the future, but I don't yet know how. If all code chunks are from the same language (e.g. python), the answer would be obvious. If a document contains multiple types of code chunks, it will be tricky.

From what you are saying above, I feel you would be okay with each python code chunk being written into a separate `.py` file, and the `purl()` output (which is an R script) will call `reticulate::source_python()` to run these `.py` files?

Thanks!

**Comment by GitHunter0** [comment]
Actually, handling the most simple case would suffice to me, i.e., take **one .Rmd** file with **only Python** chunks and convert it into **one single .py** file (so that I can call `reticulate::source_python()`)

Thanks for the feedback!

**Comment by GitHunter0** [comment]
Closing as it is being tracked by https://github.com/yihui/knitr/issues/2209

**Comment by cderv** [comment]
Oops sorry I missed this. Let's keep this one.

**Comment by GitHunter0** [comment]
In the conversion to .py files, my suggestion is to mark python chunks with `#%%` , since this is interpreted as code cells by Spyder and VScode IDEs.

**Comment by yihui** [comment]
Do you mean something like this?

```py
#%% label, chunk options ----------
code
```

I looked at the code and thought a little about the implementation yesterday. The tricky part is that I need to check if all code chunks are python beforehand. This isn't hard for a single document. For a document that contains child documents, **knitr** doesn't parse child documents in advance, so it's not possible to know if child documents contain other languages. I guess this edge case should be relatively rare, though.

**Comment by GitHunter0** [comment]
> Do you mean something like this?
> 
> ```python
> #%% label, chunk options ----------
> code
> ```

Yes @yihui , that's what I mean.

> I guess this edge case should be relatively rare, though.

I also believe this case should be rare, I would not worry much about that.

**Comment by cderv** [comment]
Could we limit the feature to document that does not use child document ? 

By emitting an error if child document is detected and explaining the function does not support it ? 

I agree this should be rare!

**Comment by GitHunter0** [comment]
We could also allow files that use child documents but throw a warning message explaining the limitation every time a child is detected.

**Comment by yihui** [comment]
Thanks for the suggestion, @GitHunter0! This is now implemented in #2503 and will be in the next release.

**How it works:** when the first code chunk of a document is not R but another language, `purl()` tangles the whole document to a script in that language instead of an `.R` script. The target language is decided by the first chunk, which keeps things simple and avoids having to pre-scan child documents.

- **Output file extension** defaults to the engine name (e.g. `python` → `.py`, `ruby` → `.rb`, `bash` → `.sh`). If the default guess isn't what you want, pass the `output` argument explicitly, e.g. `purl("foo.Rmd", "foo.py")`.
- **Chunk headers** become `#%%` code cells, using the language's comment character (so it's `#%%` for Python, and block-comment languages like C get a closed `/*%% ... */`). Editors such as Spyder and VS Code recognize `#%%` as cell separators.
- **Chunks of other languages are ignored**, so the resulting script contains only code of the target language and is directly runnable.

For your original use case, a Python-only R Markdown document:

````markdown
```{python}
import os
```

```{python}
x = 1
print(x)
```
````

`purl("foo.Rmd")` now produces `foo.py`:

```python
#%% ----
import os

#%% ----
x = 1
print(x)
```

which you can run however you like (e.g. `reticulate::source_python("foo.py")`, or directly with Python).
